### PR TITLE
Add explain query plan feature

### PR DESCRIPTION
**Pull request by qingxinhome**
**What type of PR is this?**

- [ ] API-change
- [ ] BUG
- [ ] Improvement
- [ ] Documentation
- [x] Feature
- [ ] Test and CI
- [ ] Code Refactoring

**Which issue(s) this PR fixes:**

issue #2363 

**What this PR does / why we need it:**
Realize the feature of explaining query plans

**Special notes for your reviewer:**

Not Available

**Additional documentation (e.g. design docs, usage docs, etc.):**

Not Available

### CODE REVIEW

File: pkg/sql/plan2/explain/build_query_explain.go [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+func NewExplainQueryImpl(query *plan.Query) *ExplainQueryImpl {
+	return &ExplainQueryImpl{
+		QueryPlan: query,
+	}
+}
+
+func (e *ExplainQueryImpl) ExplainPlan(buffer *ExplainDataBuffer, options *ExplainOptions) {
+	var Nodes []*plan.Node = e.QueryPlan.Nodes
+	for index, rootNodeId := range e.QueryPlan.Steps {
+		fmt.Printf("-----------------------------Printing the %v-th plan tree -----------------------\n", index)
```
Comment (daviszhen) [inline pkg/sql/plan2/explain/build_query_explain.go:39]: use logutil

File: pkg/sql/plan2/explain/build_query_explain.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+		settings := FormatSettings{
+			buffer:      buffer,
+			offset:      0,
+			indent:      4,
+			indent_char: ' ',
+			level:       0,
+		}
+
+		stack := NewStack()
+		stack.Push(&Frame{
```
Comment (daviszhen) [inline pkg/sql/plan2/explain/build_query_explain.go:49]: It is a level traverse. It really needs the stack ?

File: pkg/sql/plan2/explain/build_node_explain.go [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+	Node *plan.Node
+}
+
+func NewNodeDescriptionImpl(node *plan.Node) *NodeDescribeImpl {
+	return &NodeDescribeImpl{
+		Node: node,
+	}
+}
+
+func (ndesc NodeDescribeImpl) GetNodeBasicInfo(options *ExplainOptions) string {
```
Comment (daviszhen) [inline pkg/sql/plan2/explain/build_node_explain.go:31]: pointer?

**Review by iamlinjunhong — APPROVED** [review]

File: pkg/sql/plan2/explain/types.go [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+	indent_char byte
+	level       int
+}
+
+func NewFormatSettings() *FormatSettings {
+	return &FormatSettings{
+		buffer:      &ExplainDataBuffer{},
+		offset:      0,
+		indent:      2,
+		indent_char: ' ',
```
Comment (ouyuanning) [inline pkg/sql/plan2/explain/types.go:65]: suggest rename to indentChar

File: pkg/sql/plan2/explain/utils.go [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+package explain
+
+import (
+	"github.com/matrixorigin/matrixone/pkg/pb/plan"
+)
+
+type Frame struct {
+	node                 *plan.Node
+	next_child           int
```
Comment (ouyuanning) [inline pkg/sql/plan2/explain/utils.go:23]: suggest use camel-case style like nextChild

File: pkg/sql/plan2/explain/build_expr_explain.go [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+// limitations under the License.
+
+package explain
+
+import (
+	"fmt"
+	"github.com/matrixorigin/matrixone/pkg/pb/plan"
+)
+
+func DescribeExpr(expr *plan.Expr) string {
```
Comment (ouyuanning) [inline pkg/sql/plan2/explain/build_expr_explain.go:22]: I think most of your functions is private will be better.

**Review by ouyuanning — COMMENTED** [review]
1、suggest use camel-case style in Attribute definition。
2、the functions only use for you can change to private.

File: pkg/sql/plan2/explain/types.go (hunk 4 above)
Comment (qingxinhome) [reply pkg/sql/plan2/explain/types.go:65]: ok, I will use the camel-case style to rename it

File: pkg/sql/plan2/explain/utils.go (hunk 5 above)
Comment (qingxinhome) [reply pkg/sql/plan2/explain/utils.go:23]: ok, I will use the camel-case style to rename it

File: pkg/frontend/mysql_cmd_executor.go [hunk 7]
```diff
@@ -881,6 +883,61 @@ func (mce *MysqlCmdExecutor) handleAnalyzeStmt(stmt *tree.AnalyzeStmt) error {
...
+	}
+
+	for _, v := range stmt.Options {
+		if strings.EqualFold(v.Name, "VERBOSE") {
+			if strings.EqualFold(v.Value, "TRUE") || v.Value == "NULL" {
+				es.Verbose = true
+			} else if strings.EqualFold(v.Value, "FALSE") {
+				es.Verbose = false
+			} else {
+				return errors.New("111111", fmt.Sprintf("%s requires a Boolean value", v.Name))
```
[the remark this answers is not in the document]
Comment (qingxinhome) [reply pkg/frontend/mysql_cmd_executor.go:900]: I will choose an appropriate error code and modify it

File: pkg/frontend/mysql_cmd_executor.go [hunk 8]
```diff
@@ -1098,7 +1155,16 @@ func (mce *MysqlCmdExecutor) doComQuery(sql string) error {
 			if err = mce.handleAnalyzeStmt(st); err != nil {
 				return err
 			}
-
+		case *tree.ExplainStmt:
+			fmt.Println("Wuxiliang----statement is explain---------------->")
```
[the remark this answers is not in the document]
Comment (qingxinhome) [reply pkg/frontend/mysql_cmd_executor.go:1159]: I will delete my printouts

File: pkg/sql/plan2/explain/build_query_explain.go (hunk 1 above)
Comment (qingxinhome) [reply pkg/sql/plan2/explain/build_query_explain.go:39]: ok！

File: pkg/sql/plan2/explain/build_node_explain.go (hunk 3 above)
Comment (qingxinhome) [reply pkg/sql/plan2/explain/build_node_explain.go:31]: This method should use the pointer as the receiver instead

**Comment by yingfeng** [comment]
Fix the failed unit test before it's merged

File: pkg/sql/parsers/tree/explain.go [hunk 9]
```diff
@@ -58,6 +70,33 @@ func (node *ExplainStmt) Format(ctx *FmtCtx) {
...
+//			st.ColName.Format(ctx)
+//		}
+//	default:
+//		if stmt != nil {
+//			ctx.WriteByte(' ')
+//			stmt.Format(ctx)
+//		}
+//	}
+//}
+
```
Comment (fengttt) [inline pkg/sql/parsers/tree/explain.go:99]: I am not a big fan of comment out functions.  If not used, remove?   If it is for later, then put it next PR?

**Review by fengttt — COMMENTED** [review]
In general, you should not use fmt.Print -- where it goes?   Esp, if it is a server.

And be consistent/clean on style.

File: pkg/sql/plan2/explain/build_query_explain.go (hunk 2 above)
Comment (daviszhen) [reply pkg/sql/plan2/explain/build_query_explain.go:49]: Actually, it is a dfs but not the level traverse.

File: pkg/sql/plan2/explain/build_expr_explain.go (hunk 6 above)
Comment (qingxinhome) [reply pkg/sql/plan2/explain/build_expr_explain.go:22]: I will consider revising this question later

File: pkg/sql/plan2/explain/build_query_explain.go (hunk 2 above)
Comment (qingxinhome) [reply pkg/sql/plan2/explain/build_query_explain.go:49]: I use depth first search to traverse the query，make sure that the tree structure can be displayed normally

File: pkg/frontend/mysql_cmd_executor.go [hunk 10]
```diff
@@ -881,6 +883,61 @@ func (mce *MysqlCmdExecutor) handleAnalyzeStmt(stmt *tree.AnalyzeStmt) error {
...
+		if err != nil {
+			fmt.Sprintf("build Query statement error: '%v'", tree.String(stmt, dialect.MYSQL))
+			return errors.New("111111", fmt.Sprintf("Build Query statement error:'%v'", tree.String(stmt.Statement, dialect.MYSQL)))
+		}
+
+		buffer := explain.NewExplainDataBuffer(250)
+		explainQuery := explain.NewExplainQueryImpl(query)
+		explainQuery.ExplainPlan(buffer, es)
+	*/
+	return errors.New("111111", "not support explain statment now")
```
[the remark this answers is not in the document]
Comment (qingxinhome) [reply pkg/frontend/mysql_cmd_executor.go:938]: ok， I will add the explain mysql client display fetaure

File: pkg/sql/parsers/tree/explain.go (hunk 9 above)
Comment (qingxinhome) [reply pkg/sql/parsers/tree/explain.go:99]: Sorry, I'll delete the fmt printout and the comment

File: pkg/sql/plan2/explain/build_query_explain.go (hunk 2 above)
Comment (fengttt) [reply pkg/sql/plan2/explain/build_query_explain.go:49]: DFS walk should not need an explicit stack.   your recursive call stack is your node stack.

File: pkg/sql/plan2/explain/init.go [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+	SysFunsAndOperatorsMap = make(map[string]FunctionProtoType)
+	for _, fun := range BuiltinFunctions {
+		SysFunsAndOperatorsMap[fun.Name] = fun
+	}
+}
+
+type FunctionKind int32
+
+const (
+	STANDARD_FUNCTION          FunctionKind = 0
```
Comment (fengttt) [inline pkg/sql/plan2/explain/init.go:31]: What do you need all these?    We have too many function files already, explain should not need this, ask mingsong or yuanning, work it out in one place.

Comment (qingxinhome) [reply pkg/sql/plan2/explain/init.go]: OK , Then I'll change it in the functions.go file

File: pkg/sql/plan2/explain/build_query_explain.go (hunk 2 above)
Comment (qingxinhome) [reply pkg/sql/plan2/explain/build_query_explain.go]: Yes, I have traversed the plan tree recursively

File: pkg/sql/plan2/build.go [hunk 12]
```diff
@@ -35,6 +34,18 @@ func buildPlan(ctx CompilerContext, stmt tree.Statement) (*Query, error) {
 	return query, nil
 }
 
+func BuildPlan2(ctx CompilerContext, stmt tree.Statement) (*plan.Query, error) {
```
Comment (ouyuanning) [inline pkg/sql/plan2/build.go:37]: I think you can move /explain/buildexplain_test.go to  /plan2/    

then there will be not necessary to  add BuildPlan2() and NewMockOptimizer2()

Comment (ouyuanning) [inline pkg/sql/plan2/build.go:37]: you can move /plan2/explain/buildexplain_test.go to /plan2/
then you can remove BuildPlan2()  and NewMockOptimizer2（）

File: pkg/sql/plan/types.go [hunk 13]
```diff
@@ -536,6 +545,25 @@ func (d ShowCreateDatabase) ResultColumns() []*Attribute {
 	return attrs
 }
 
+func (e ExplainQuery) String() string {
```
Comment (ouyuanning) [inline pkg/sql/plan/types.go:548]: I don't think it's a good way to write impl methods in types.go

**Review by ouyuanning — COMMENTED** [review]
you can explain your code in Monday's meeting
i think you may have a talk with Dr Tian for this PR.

**Review by ouyuanning — APPROVED** [review]

**Pull request merged**